src/storage-client/src/controller.rs

```diff
@@ -374,9 +375,13 @@ pub trait StorageController: Debug + Send {
         commands: Vec<(GlobalId, Vec<TimestamplessUpdate>)>,
     ) -> Result<tokio::sync::oneshot::Receiver<Result<(), StorageError>>, StorageError>;
 
-    /// Returns a [`MonotonicAppender`] which is a oneshot-esque struct that can be used to
-    /// monotonically append to the specified [`GlobalId`].
-    fn monotonic_appender(&self, id: GlobalId) -> Result<MonotonicAppender, StorageError>;
+    /// Returns a [`MonotonicAppender`] which can be used to monotonically append to the specified
+    /// [`GlobalId`].
+    fn monotonic_appender(
+        &self,
+        id: GlobalId,
+        catalog_revision: Arc<AtomicU64>,
```

Comment:
Why does the storage controller need to care about a catalog revision? Looks like a API boundary violation but maybe I'm missing something

Reply:
Having thought about this more since last week I agree with Petros. I have a vague idea but I'm unfamiliar with enough parts it's not worth writing down.

Reply:
Sounds much cleaner, thanks!

Reply:
Just updated the PR, I did need to add a new API to the `StorageController` trait called `invalidate_monotonic_appenders` which closes the channel as I described above, let me know what you think!

Reply:
Hm, I'm still confused as to why we need this. What is actually changing that makes the previous appender invalid? Also, can't we simply drop the currently active appender that we don't want to use anymore and request a new one?

Reply:
What this specific problem boils down to is how do we invalidate the cache of appenders that the Adapters use?

The first approach used the Catalog transient revision, if the Catalog changed then there's a chance the webhook changed, so we should invalidate the cache. This required the `MonotonicAppender`s to know about the Catalog to realize they were invalid, which I agree maybe isn't the cleanest approach.

This second attempt is to invalidate the cache by closing the `MonotonicAppender`/channel so when an Adapter attempts to send data, it will fail. This is why I introduced the `invalidate_monotonic_appenders(...)` method, because the only way to close the channel is to drop the `Receiver`, which is owned by the spawned task in `collection_mgmt`. I agree this feels a bit verbose/unnecessary, but was less typing than the third option.

A third option is `MonotonicAppender` grows the ability to be closed, regardless of whether or not the underlying channel closed. For example, something like:

```
struct MonotonicAppender {
  tx: mpsc::Sender<...>,
  closed: Arc<AtomicBool>,
}

struct MonotonicAppenderHandle {
  closed: Arc<AtomicBool>,
}

impl MonotonicAppenderHandle {
  fn close(self) {
    self.closed.store(true);
  }
}
```

We could hold onto the `MonotonicAppenderHandle`s in the `CatalogState`, and replace the current call to `invalidate_monotonic_appender(...)`, with calling `.close()` on any handles we have. I opted not to take this approach because it initially seemed more involved, but maybe it's the only one that doesn't add some odd dependency or tech debt?

What do y'all think @mjibson and @petrosagg? Also thanks for iterating on this with me!

Reply:
I like the approach in the current PR (but also I don't know anything about the storage controller).

Reply:
Ok, I think I'm up to date as to why this is needed. When source `a` gets renamed to `b` we don't want any more requests arriving at `/a` to affect source `b`. However, this is an entirely adapter concern, the storage controller knows nothing about catalogs or names or renames and so I don't think there is a place for an invalidate API at all.

The way I'd go about it is that adapter requests a `MonotonicAppender` for a `GlobalId` and then makes sure that this appender is used in the right places. Each URL handler could hold onto an `Arc<Mutex<MonotonicAppender>>` that the adapter can swap when a rename happens. This type effectively encodes that there is a mutable connection between a URL handler and the associated `GlobalId`.

Reply:
> When source a gets renamed to b we don't want any more requests arriving at /a to affect source b.

Exactly! Another wrinkle though is soon/eventually the Adapter and Coordinator will be in two separate processes and only communicate over the network, so we don't want to introduce non-serializable dependencies like an `Arc<Mutex<...>>` between the two. To future proof ourselves, the approach we're aiming for is something like:

1. Adapter makes request to append webhook
2. Request fails
3. Adapter has to fetch a new resource

This is why we've pushed the invalidation logic into the channel, because whether it's a channel today or a TCP connection in the future, this approach should work either way.

Reply:
> eventually the Adapter and Coordinator will be in two separate processes and only communicate over the network

I think this actually might end up not being the case. At the least, I wouldn't design anything today based on this restriction. If everything is going to live in clusterd, there's no hard requirement that the communication is serializable.

Reply:
@petrosagg and I chatted this morning and we're aligned, I just pushed another update to the PR which takes approach 3 that I described above.

Instead of asking the `storage-controller` to invalidate the returned channel, we wrap a [`MontonicAppender`] in a new [`WebhookAppender`]. The new [`WebhookAppender`] holds an `Arc<AtomicBool>` (with some newtype wrapping) which the Coordinator also holds a clone of. When a webhook source (or its parent schema) gets altered the Coordinator on its own invalidates the outstanding [`WebhookAppender`]s.

I also updated the Coordinator consistency checks to make sure the `active_webhooks` map is kept in sync with the Catalog and will kick off a nightly run.